Add integration for iotty Smart Home

pburgio opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is the first implementation of Hass support to iotty Smart Devices

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29624

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

pburgio committed:
Initial import 0.0.2

pburgio committed:
Initial import 0.0.2

pburgio committed:
Initial import 0.0.2

pburgio committed:
Initial import 0.0.2

pburgio committed:
Fixes to URL, and removed commits

pburgio committed:
Fixes to URL, and removed commits

pburgio committed:
Fixes to URL, and removed commits

pburgio committed:
Fixes to URL, and removed commits

pburgio committed:
Added first test for iotty

pburgio committed:
Added first test for iotty

pburgio committed:
Added first test for iotty

pburgio committed:
Fixed conflicts

pburgio committed:
First release

pburgio committed:
First release

pburgio committed:
First release

pburgio committed:
Merge branch 'iotty' of https://iotty.visualstudio.com/iotty/_git/homeassistant.core into iotty

jpbede commented:
Please also add the integration to the [documentation](https://github.com/home-assistant/home-assistant.io) and link it here.

pburgio commented:
> Please also add the integration to the [documentation](https://github.com/home-assistant/home-assistant.io) and link it here.

Hello, our documentation is ready, but I don't have rights to access the repo. How can I issue a pull request? Shall I use a fork?

pburgio committed:
Merge branch 'dev' into iotty

edenhaus commented on homeassistant/components/iotty/__init__.py:
```diff
@@ -0,0 +1,55 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+# Uncomment these when developing
+# _LOGGER.setLevel(logging.DEBUG)
+# logging.getLogger("iottycloud").setLevel(logging.DEBUG)
```
```suggestion
```

Please remove any commented out code.
n this specific case, the logger can be set to debug via https://www.home-assistant.io/integrations/logger/

pburgio replied:
Thanks! Will next commits of this PR

edenhaus commented on homeassistant/components/iotty/__init__.py:
```diff
@@ -0,0 +1,55 @@
+# _LOGGER.setLevel(logging.DEBUG)
+# logging.getLogger("iottycloud").setLevel(logging.DEBUG)
+
+# For your initial PR, limit it to 1 platform.
```
```suggestion
```

edenhaus commented on homeassistant/components/iotty/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "config_flow": true,
+  "dependencies": ["application_credentials"],
+  "documentation": "https://www.home-assistant.io/integrations/iotty",
+  "homekit": {},
```
Remove empty keys

pburgio replied:
Done.

edenhaus commented on homeassistant/components/iotty/__init__.py:
```diff
@@ -0,0 +1,55 @@
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = iotty
+
+    await asyncio.sleep(3)
```
Why do you need this?

edenhaus commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,92 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class IottyLightSwitch(SwitchEntity):
```
For new integrations, we require the following:
- Support for [the new entity naming](https://developers.home-assistant.io/docs/core/entity#has_entity_name-true-mandatory-for-new-integrations)
- Services must be [translatable](https://developers.home-assistant.io/blog/2023/07/11/translating-services)
- Entities and their attributes should be [translatable](https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations/)

Please adopt your code to add support for it.

pburgio replied:
Thanks. Will include it in next commits.

pburgio commented:
> Please also add the integration to the [documentation](https://github.com/home-assistant/home-assistant.io) and link it here.

Opened a PR for that. All checks passed. Waiting for approval
[link](https://github.com/home-assistant/home-assistant.io/pull/29624)

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Reviewers request #1
- Removed clutter
- Added support for new naming convention for IottySmartSwitch entity

pburgio committed:
Reviewers request #1
- Removed clutter
- Added support for new naming convention for IottySmartSwitch entity

pburgio committed:
Reviewers request #1
- Removed clutter
- Added support for new naming convention for IottySmartSwitch entity

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'iotty' of https://github.com/pburgio/homeassistant.core into iotty

pburgio committed:
Removed commmented code

pburgio committed:
Removed commmented code

pburgio committed:
Removed commmented code

pburgio committed:
Some modifications

pburgio committed:
Some modifications

pburgio committed:
Some modifications

pburgio committed:
Some modifications

pburgio committed:
Modified REST EP for iotty CloudApi

pburgio committed:
Modified REST EP for iotty CloudApi

pburgio committed:
Modified REST EP for iotty CloudApi

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Rebased and resolved conflicts

pburgio committed:
Rebased and resolved conflicts

pburgio committed:
Rebased and resolved conflicts

pburgio committed:
Removed empty entries in manifest.json

pburgio committed:
Removed empty entries in manifest.json

pburgio committed:
Added test_config_flow

pburgio committed:
Added test_config_flow

pburgio committed:
Resolved conflicts

pburgio committed:
Merge branch 'home-assistant:dev' into iotty

edenhaus commented on homeassistant/components/iotty/api.py:
```diff
@@ -0,0 +1,108 @@
+        _LOGGER.debug("store_entity %s", device_id)
+        self._entities[device_id] = entity
+
+    async def _polling(self) -> None:
```
Can we use a Coordinator instead?

pburgio replied:
We can certainly. Is there a specific reason to do so?

pburgio replied:
Anyway, will address this in next commits

edenhaus replied:
The coordinator does the scheduling and co already and we should reuse it instead of creating all by our own. Please add a coordinator

Also if the user disable a certain device, it looks like this API is still polling the information about that device, which can be omitted by using a coordinator

pburgio replied:
Ok, implemented coordinator

pburgio commented on homeassistant/components/iotty/__init__.py:
```diff
@@ -0,0 +1,48 @@
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = iotty
+
+    #    await asyncio.sleep(3)
+    #    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```
Just clutter. I will remove it and resubmit. Sorry.

pburgio commented on homeassistant/components/iotty/.gitignore:
```diff
@@ -0,0 +1,3 @@
+.vscode/*
+
```
Old stuff. I can remove it, if you prefer

pburgio commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,93 @@
+
+    _attr_entity_category = EntityCategory.CONFIG
+    _attr_device_class = SwitchDeviceClass.SWITCH
+    _iotty_cloud: Any
+    _iotty_device: Any
```
Will fix in next commits

pburgio commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,93 @@
+    iotty = hass.data[DOMAIN][config_entry.entry_id]
+    _ls_list = await iotty.devices(LS_DEVICE_TYPE_UID)
+
+    _LOGGER.info("Found %d LightSwitches", len(_ls_list))
```
Will fix in next Commit

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Fix as requested by @edenhaus

pburgio committed:
Fix as requested by @edenhaus

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'home-assistant:dev' into iotty

pburgio committed:
Merge branch 'iotty' of https://github.com/pburgio/homeassistant.core into iotty

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'iotty' of https://github.com/pburgio/homeassistant.core into iotty

pburgio committed:
Added test_init

pburgio committed:
Added test_init

pburgio committed:
Removed comments, added one assert

pburgio committed:
Removed comments, added one assert

pburgio committed:
Added TEST_CONFIG_FLOW

pburgio committed:
Added TEST_CONFIG_FLOW

pburgio committed:
Added test for STORE_ENTITY

pburgio committed:
Added test for STORE_ENTITY

pburgio committed:
Increased code coverage

pburgio committed:
Increased code coverage

pburgio committed:
Full coverage for api.py

pburgio committed:
Full coverage for api.py

pburgio committed:
Added tests for switch component

pburgio committed:
Added tests for switch component

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'iotty' of https://github.com/pburgio/homeassistant.core into iotty

pburgio committed:
Converted INFO logs onto DEBUG logs

pburgio committed:
Converted INFO logs onto DEBUG logs

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'iotty' of https://github.com/pburgio/homeassistant.core into iotty

pburgio commented:
Hello @edenhaus it's ready for a second round of review. Considering our architecture, We think that a Coordinator might be an overkill, and did not use it. If you want, we can discuss this

Thanks

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'dev' into iotty

pburgio commented:
> Hello @edenhaus it's ready for a second round of review. Considering our architecture, We think that a Coordinator might be an overkill, and did not use it. If you want, we can discuss this
> 
> Thanks

Hello,

do you by chance had time to look at this? Is there any chance we can be part of the release in January?

Best
Paolo

pburgio committed:
Merge branch 'dev' into iotty

edenhaus commented on homeassistant/components/iotty/.gitignore:
```diff
@@ -0,0 +1 @@
+homeassistant/components/iotty/.vscode/*
```
This file needs to be removed

pburgio replied:
Ack

edenhaus commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,110 @@
+        return self._iotty_device.device_id
+
+    @property
+    def name(self) -> str:
```
Above, you set `_attr_name=None` to use the device name. Why are you creating this property?

pburgio replied:
Clutter. Will remove it

edenhaus commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,110 @@
+        """Initialize the LightSwitch device."""
+        super().__init__()
+
+        if iotty_cloud is None:
+            raise ValueError("iotty_cloud")
```
```suggestion
```

Can never happen

edenhaus commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,110 @@
+    hass_data = hass.data[DOMAIN]
+
+    iotty = hass_data[config_entry.entry_id]
+    if iotty is None:
+        _LOGGER.error("Cannot retrieve iotty MW component")
+        raise ValueError("iotty")
```
```suggestion
```

Can never happen as you set it in the init file

pburgio replied:
Ack

edenhaus commented on homeassistant/components/iotty/api.py:
```diff
@@ -0,0 +1,113 @@
+                    else:
+                        self._entities[device.device_id].schedule_update_ha_state()
+
+            await asyncio.sleep(5)
```
Why is the sleep required?

pburgio replied:
Removed

edenhaus commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,110 @@
+        if iotty_cloud is None:
+            raise ValueError("iotty_cloud")
+
+        if iotty_device is None:
```
Can this ever happen?

pburgio replied:
Lines 35-36, were already removed to address another comment. Do you mean line 38?

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Merge branch 'home-assistant:dev' into iotty

pburgio committed:
Removed .gitignore from commits

pburgio committed:
Removed .gitignore from commits

pburgio committed:
Modifications to SWITCH.PY

pburgio committed:
Modifications to SWITCH.PY

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Fixed tests for SWITCH

pburgio committed:
Resolved conflicts

pburgio committed:
First working implementation of Coordinator

pburgio committed:
Increased code coverage

pburgio committed:
Full code coverage

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Missing a line in testing

edenhaus commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+
+        await super().async_config_entry_first_refresh()
+
+        await self._hass.config_entries.async_forward_entry_setups(
+            self.config_entry, PLATFORMS
+        )
```
This should be moved to `async_setup_entry`

edenhaus commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+        for device in self._devices:
+            res = await self.iotty.get_status(device.device_id)
+
+            if RESULT not in res or STATUS not in res[RESULT]:
```
```suggestion
            if not (status := res.get(RESULT, {}).get(STATUS)):
```

pburgio replied:
Thanks

edenhaus commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+            if RESULT not in res or STATUS not in res[RESULT]:
+                _LOGGER.warning("Unable to read status for device %s", device.device_id)
+            else:
+                status = res[RESULT][STATUS]
```
```suggestion
```

edenhaus commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+        _LOGGER.debug("Fetching devices status from iottyCloud")
+
+        for device in self._devices:
+            res = await self.iotty.get_status(device.device_id)
```
As you make one request per device, I suggest creating a coordinator per device.
So you don't need to override `async_config_entry_first_refresh` to fetch all devices

shapournemati-iotty replied:
Hi, I'm not sure if that's what you meant, but I have used the coordinator entity! Let me know if that satisfies your request

edenhaus commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+                    "Retrieved status: '%s' for device %s", status, device.device_id
+                )
+                device.update_status(status)
+                if device.device_id in self._entities:
```
If one coordinator per device, this can be omitted as the entities will subscribe to updates if the are enabled

shapournemati-iotty replied:
Hi, I'm not sure if that's what you meant, but I have used the coordinator entity! Let me know if that satisfies your request

edenhaus commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,99 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class IottyLightSwitch(SwitchEntity):
```
```suggestion
class IottyLightSwitch(SwitchEntity, CoordinatorEntity[Device]):
```

Please adjust the entity

edenhaus commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,99 @@
+        if d.device_type == LS_DEVICE_TYPE_UID
+    ]
+    _LOGGER.debug("Found %d LightSwitches", len(entities))
+    for e in entities:
+        coordinator.store_entity(e.device_id, e)
```
```suggestion
```

Can be removed by using a CoordinatorEntity

shapournemati-iotty replied:
Done! Let me know if that satisfies your request

pburgio committed:
Merge branch 'dev' into iotty

pburgio committed:
Update homeassistant/components/iotty/__init__.py

Co-authored-by: Robert Resch <robert@resch.dev>

pburgio committed:
Update homeassistant/components/iotty/__init__.py

Co-authored-by: Robert Resch <robert@resch.dev>

pburgio commented on homeassistant/components/iotty/const.py:
```diff
@@ -0,0 +1,16 @@
+
+
+LOGGER = logging.getLogger(__package__)
+UPDATE_INTERVAL = timedelta(seconds=10)
```
I checked a bit what the other integrations are doing, and I found it as a reasonable value. It can certainly be bigger. Do you have any recommendations on this?

shapournemati-iotty replied:
Hi, I have changed this to be 30s, I hope this is fine

pburgio commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+        """Initialize the coordinator."""
+        _LOGGER.debug("Initializing iotty data update coordinator")
+        self.config_entry = entry
+        self._hass = hass
```
Will do this, thanks

pburgio committed:
Modified coordinator as per request by edenhaus

pburgio committed:
Merge branch 'dev' into iotty

shapournemati-iotty committed:
use coordinator entities for switches

shapournemati-iotty commented:
> > Please also add the integration to the [documentation](https://github.com/home-assistant/home-assistant.io) and link it here.
> 
> Opened a PR for that. All checks passed. Waiting for approval [link](https://github.com/home-assistant/home-assistant.io/pull/29624)

@edenhaus any chance we can get an approval on this one as well? Thanks for your patience!

shapournemati-iotty committed:
move platforms to constants

shapournemati-iotty commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,97 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
```
Moved to const.py as requested

edenhaus commented:
The PR is still in draft and, therefore, not ready for another review.
Please mark the PR ready for review and the bot will ping me. No need to ping me manually.

pburgio committed:
Merge branch 'dev' into iotty

shapournemati-iotty committed:
fix whitespace with ruff-format

emontnemery committed:
Merge branch 'dev' into iotty

emontnemery reviewed (changes requested):
CI is failing, please hit the ready for review button when that's been addressed.

shapournemati-iotty committed:
correct iotty entry in application_credentials list

pburgio committed:
Merge branch 'dev' into iotty

emontnemery commented on homeassistant/components/iotty/api.py:
```diff
@@ -0,0 +1,49 @@
+class IottyProxy(CloudApi):
+    """Provide iotty authentication tied to an OAuth2 based config entry."""
+
+    # We shall remove these
+    _devices: list[Device]
+    _entities: dict
+    _coroutine: Any
```
Either remove these if they are not needed, or remove the comment if they are needed.

shapournemati-iotty replied:
That was indeed some code that was got lost in the transition, thanks for the heads-up!

emontnemery replied:
They are still not removed though

emontnemery commented on homeassistant/components/iotty/application_credentials.py:
```diff
@@ -0,0 +1,14 @@
+"""application_credentials platform the iotty integration."""
```
```suggestion
"""Application credentials platform for iotty."""
```

emontnemery commented on homeassistant/components/iotty/const.py:
```diff
@@ -0,0 +1,19 @@
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+
+LOGGER = logging.getLogger(__package__)
+UPDATE_INTERVAL = timedelta(seconds=30)
```
Please move this constant, which is only used in `coordinator.py` to `coordinator.py`.

Same comment for other constants which are only used once.

shapournemati-iotty replied:
Hi, the way you are suggesting is the way it was originally. During the review, @edenhaus actually suggested moving it into the consts, and so we did.

emontnemery replied:
Constants which are used only in one module should not be in const.py. Constants which are used more than once should be in const.py, I just double checked and @edenhaus agrees. Maybe there was some misunderstanding.

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+            hass, aiohttp_client.async_get_clientsession(hass), session
+        )
+
+    def store_entity(self, device_id: str, entity: Entity) -> None:
```
I don't like the name of this method because it suggests permanent store. How about `add_entity` or `set_entity` instead?

shapournemati-iotty replied:
I agree, the first thought was that it would actually be stored for a long time, but since it is not actually permanent, I have adopted the naming you suggested.

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+
+        await super().async_config_entry_first_refresh()
+
+        await self.hass.config_entries.async_forward_entry_setups(
+            self.config_entry, PLATFORMS
+        )
```
In a previous review, @edenhaus requested to move this to `async_setup_entry`, that comment was marked as resolved, but the code is still here. Either fix it or explain why that's not possible.

shapournemati-iotty replied:
This as well was either lost in a merge, or just not commited. Now it should be solved.

emontnemery replied:
Still not fixed

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+    async def async_config_entry_first_refresh(self) -> None:
+        """Override the first refresh to also fetch iotty devices list."""
+        _LOGGER.debug("Fetching devices list from iottyCloud")
+        self._devices = await self.iotty.get_devices()
```
Can this happen after the super() call? Also, why do we only get the device list once, don't we need to do that at every update and handle new devices?

emontnemery replied:
You need to fix it such that new devices are automatically added, here's an example of how it can be implemented: https://github.com/home-assistant/core/blob/e472af6fb453784702dc5510dd02ce5c7257a231/homeassistant/components/meater/sensor.py#L146-L174

About removed devices, it depends on if the user needs to manually remove already installed devices from the iotty supporting app or not. If they do, and already installed devices are always included in the API responses regardless of if they lose power or can't connect to the WiFi network you should automatically clean up such devices by adding another listener to the coordinator which removed devices removed by the user.

If devices are not included in API responses when they lose power, you should instead allow the user to remove devices by implementing a function like this in `__init__.py`
```py
async def async_remove_config_entry_device(
    hass: HomeAssistant, config_entry: ConfigEntry, device_entry: DeviceEntry
) -> bool:
```

shapournemati-iotty replied:
Hi, I have followed your suggestions, let me know if that is ok and satisfies your request

emontnemery replied:
It looks good, except that you're not handling removed devices.

shapournemati-iotty replied:
Hi, in my case the API will not return any removed device, so according to your previous comment (https://github.com/home-assistant/core/pull/103073#discussion_r1613504099), I should be in this situation:

> If devices are not included in API responses when they lose power, you should instead allow the user to remove devices by implementing a function like this in __init__.py

This is what I did. The function is empty because I do not need to perform further action, since the API is already behaving as the device does not exist, so the only thing left is removing the device from the entity.
According to the official docs (https://developers.home-assistant.io/docs/device_registry_index?_highlight=async_remove_config_entry_device#removing-devices) it should be done automatically by hass.

> When the user clicks the delete device button for the device and confirms it, async_remove_config_entry_device will be awaited and if True is returned, the config entry will be removed from the device. If it was the only config entry of the device, the device will be removed from the device registry.

Looks to me like I did all of the things requested by you and the documentation, am I missing something? What kind of other cleanup am I supposed to do?

emontnemery replied:
I'm confused, you write:

> Hi, in my case the API will not return any removed device, so according to your previous comment (https://github.com/home-assistant/core/pull/103073#discussion_r1613504099), I should be in this situation:

>> If devices are not included in API responses when they lose power, you should instead allow the user to remove devices by implementing a function like this in __init__.py

Let me try to rephrase:

If a device, which has previously been included in API responses, is disconnected from power but the user takes no action to remove it via the iotty app, will that device be included in or omitted from API-responses?

emontnemery replied:
This was discussed and resolved in a Discord chat

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+        _LOGGER.debug("Fetching devices status from iottyCloud")
+
+        for device in self._devices:
+            res = await self.iotty.get_status(device.device_id)
```
What @edenhaus meant is that it seems the coordinator is not useful, because you make one API call for each device. Instead, each device could have its own coordinator.
I think the direction here depends a bit on the plan for future PR which adds functionality, where it may be more useful to keep all data in a single coordinator.

shapournemati-iotty replied:
Thanks for pointing that out. Actually, as you already said at the end, our future plans would actually benefit from having all the info in one coordinator. Moreover, we are also considering using a single API call to get all the devices info at once, and then cycle through them to get the updated info, which cannot be done using one coordinator per device.

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+
+        for device in self._devices:
+            res = await self.iotty.get_status(device.device_id)
+            json = res.get(RESULT, {})
+            if not (hasattr(json, "get") and callable(json.get)) or not (
+                status := json.get(STATUS)
+            ):
```
The library call to `self.iotty.get_status` returns decoded JSON data, which I think we try to validate here. I can't follow the logic though.
If we don't trust the returned data is valid, we should check the types, maybe something like this:

```py
    if (
        not isinstance(res, dict)
        or RESULT not in res
        or not isinstance(json := res[RESULT], dict)
        or not (status := json.get(STATUS))
    ):
        _LOGGER.warning("Unable to read status for device %s", device.device_id)
        continue
```

shapournemati-iotty replied:
I see what you mean. To the eye of the implementer, the previous code didn't seem to cryptic, but with the proposed solution it is more clear (and more complete), so it has been used.

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+    """Class to manage fetching Iotty data."""
+
+    config_entry: ConfigEntry
+    _entities: dict
```
Please annotate this dict

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,122 @@
+"""Implement a iotty Light Switch Device."""
+
```
Import annotations from __future__, in this file and in all other files
```suggestion
from __future__ import annotations

```

emontnemery replied:
This has not been fixed. Please add it to all files to avoid inconsistent behavior by the type checker.

shapournemati-iotty replied:
I didn't notice you asked to do that for all files, now it is done.

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,122 @@
+        self.async_write_ha_state()
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Activate the iotty LightSwitch component."""
+    _LOGGER.debug("Setup SWITCH entry id is %s", config_entry.entry_id)
+
+    hass_data = hass.data[DOMAIN]
+
+    coordinator: IottyDataUpdateCoordinator = hass_data[config_entry.entry_id]
+
+    entities = [
+        IottyLightSwitch(
+            coordinator=coordinator, iotty_cloud=coordinator.iotty, iotty_device=d
+        )
+        for d in coordinator.data.devices
+        if d.device_type == LS_DEVICE_TYPE_UID
+    ]
+    _LOGGER.debug("Found %d LightSwitches", len(entities))
+    for e in entities:
+        coordinator.store_entity(e.device_id, e)
+
+    async_add_entities(entities)
```
Please move this to before the entity class

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,122 @@
+        if d.device_type == LS_DEVICE_TYPE_UID
+    ]
+    _LOGGER.debug("Found %d LightSwitches", len(entities))
+    for e in entities:
+        coordinator.store_entity(e.device_id, e)
```
You said in a previous review comment this would be removed, that doesn't seem to be done though?

shapournemati-iotty replied:
I think it might got lost in the merge, or maybe just forgot to commit it previously. It got fixed with the last commit, thanks!

emontnemery reviewed (changes requested):
Some additional review comments. I've not reviewed the tests.

shapournemati-iotty committed:
minor style improvements

shapournemati-iotty committed:
refactor function name

pburgio committed:
Merge branch 'dev' into iotty

emontnemery commented:
There are unresolved review comments, please address them, then hit the "Ready for review"-button

shapournemati-iotty committed:
handle new and deleted devices

pburgio committed:
Merge branch 'dev' into iotty

emontnemery commented on homeassistant/components/iotty/__init__.py:
```diff
@@ -0,0 +1,57 @@
+        hass, entry, session
+    )
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data_update_coordinator
+    hass.data[DOMAIN].setdefault(KNOWN_DEVICES, set())
```
Please change to use this pattern instead of storing untyped data in `hass.data` https://developers.home-assistant.io/blog/2024/04/30/store-runtime-data-inside-config-entry

shapournemati-iotty replied:
Hi, I've followed the instructions, I hope that satisfies your request

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,165 @@
+
+        devices = coordinator.data.devices
+        entities = []
+        known_devices: set = hass.data[DOMAIN][KNOWN_DEVICES]
```
Why do we store this in `hass.data`, it's only used by switch to keep track of which devices it's already seen.

shapournemati-iotty replied:
Hi, since it was handled that way in the example you provided me I thought that was a good practice.
If you are not satisfied with that, could I please ask you what you recommend, since the documentation does not cover this topic?
Since it is a function outside of a class (hence, I cannot use fields), am I supposed to work with global variables, other storages, or something else?

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,165 @@
+                        iotty_cloud=coordinator.iotty,
+                        iotty_device=d,
+                    )
+                    for d in coordinator.data.devices
+                    if d.device_type == LS_DEVICE_TYPE_UID
+                    and d.device_id == device.device_id
```
What's up with the looping here? The outer loop already loops over `coordinator.data.devices`, why do we loop over it *again*?

shapournemati-iotty replied:
My bad, it's fixed now

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,110 @@
+        """Fetch data from iottyCloud device."""
+        _LOGGER.debug("Fetching devices status from iottyCloud")
+
+        all_devices = await self.iotty.get_devices()
+        new_devices = [
+            d
+            for d in all_devices
+            if not any(x.device_id == d.device_id for x in self._devices)
+        ]
+        removed_devices = [
+            d
+            for d in self._devices
+            if not any(x.device_id == d.device_id for x in all_devices)
+        ]
+
+        for new_device in new_devices:
+            self._devices.append(new_device)
+
+        for removed_device in removed_devices:
+            self._devices.remove(removed_device)
```
What's the point of all this when we don't seem to care about either the `nex_devices` or `removed_devices` lists?
Wouldn't this be the same thing:
```
self._devices = await self.iotty.get_devices()
```

shapournemati-iotty replied:
Good point, I have changed it just as you suggested

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,110 @@
+        """Fetch data from iottyCloud device."""
+        _LOGGER.debug("Fetching devices status from iottyCloud")
+
+        all_devices = await self.iotty.get_devices()
+        new_devices = [
+            d
+            for d in all_devices
+            if not any(x.device_id == d.device_id for x in self._devices)
+        ]
+        removed_devices = [
+            d
+            for d in self._devices
+            if not any(x.device_id == d.device_id for x in all_devices)
+        ]
+
+        for new_device in new_devices:
+            self._devices.append(new_device)
+
+        for removed_device in removed_devices:
+            self._devices.remove(removed_device)
```
Why can we use `list.remove` here to remove list items but when building the list of removed items we had to do this:
```
        removed_devices = [
            d
            for d in self._devices
            if not any(x.device_id == d.device_id for x in all_devices)
        ]
```

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,165 @@
+            self._iotty_device.is_on = device.is_on
+            self.async_write_ha_state()
+        except StopIteration:
+            _LOGGER.debug(
+                "Device with id [%s] is no longer handled by iotty, please remove it from hass",
+                self._iotty_device.device_id,
+            )
```
Who are we asking to remove the device?

shapournemati-iotty replied:
This is just a console error that the user might see if they decide to change the log level for the integration. Would you prefer a different wording?

emontnemery replied:
Since the integration now cleans up, deletion this is automatically done, correct? I'd suggest to just remove the `,please remove it from hass" part of the message.

emontnemery reviewed (changes requested):
A few more comments. I've still not reviewed the tests.

emontnemery commented:
Please also check and fix the lint errors 👍

shapournemati-iotty committed:
Merge branch 'dev' into iotty

shapournemati-iotty committed:
improve code for adding devices after first initialization

shapournemati-iotty committed:
use typed config entry instead of adding known devices to hass.data

pburgio committed:
Merge branch 'dev' into iotty

emontnemery commented:
I want to better understand how iotty handles devices which lose power or are for some other reason not connected to the network, please see my new comment above. You're also welcome to contact me on discord, it might be quicker.

shapournemati-iotty committed:
improve iotty entity removal

shapournemati-iotty committed:
test listeners update cycle

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,116 @@
+            if not any(x.device_id == d.device_id for x in current_devices)
+        ]
+
+        entries = er.async_entries_for_config_entry(
+            self._entity_registry, self.config_entry.entry_id
+        )
+
+        for device_to_remove in removed_devices:
+            entity_to_remove = [
+                x for x in entries if x.unique_id == device_to_remove.device_id
+            ][0]
+            self._entity_registry.async_remove(entity_to_remove.entity_id)
+
```
Instead of just creating entities, can the integration create devices?

shapournemati-iotty replied:
Done

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,170 @@
+    coordinator.async_add_listener(async_update_data)
+
+
+class IottyLightSwitch(SwitchEntity, CoordinatorEntity[Device]):
```
Should the switch entity be part of a device instead?

shapournemati-iotty replied:
Done

shapournemati-iotty committed:
handle iotty as devices and not only as entities

shapournemati-iotty committed:
Merge branch 'dev' into iotty

shapournemati-iotty committed:
fix test typing for mock config entry

emontnemery commented on homeassistant/components/iotty/__init__.py:
```diff
@@ -0,0 +1,60 @@
+        hass, entry, session
+    )
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data_update_coordinator
+    entry.runtime_data = KnownDevicesData(set())
```
Is there a reason to store a reference to the coordinator in `hass.data`, or can we store both in the config entry's runtime data?

shapournemati-iotty replied:
Hi, looking at other examples (and even in the documentation) this looks like is the standard, so that's the main reason why it's inside hass.data. Do you see any drawbacks in that?
I think it could also be in runtime_data, though it looks a bit less convenient to access the two that way

emontnemery commented on homeassistant/components/iotty/coordinator.py:
```diff
@@ -0,0 +1,112 @@
+        )
+        self._device_registry = dr.async_get(hass)
+
+    def set_entity(self, device_id: str, entity: Entity) -> None:
+        """Store iotty device within Hass entities."""
+        _LOGGER.debug("Storing device '%s' in entities", device_id)
+        self._entities[device_id] = entity
```
Is this method still needed? It seems like it's only called from tests?

shapournemati-iotty replied:
removed!

shapournemati-iotty committed:
test with fewer mocks for an integration test style opposed to the previous unit test style

shapournemati-iotty committed:
remove useless tests and add more integration style tests

shapournemati-iotty committed:
check if device_to_remove is None

shapournemati-iotty committed:
integration style tests for turning switches on and off

shapournemati-iotty committed:
remove redundant coordinator tests

shapournemati-iotty committed:
check device status after issuing command in tests

shapournemati-iotty committed:
Merge branch 'dev' into iotty

shapournemati-iotty committed:
remove unused fixtures

shapournemati-iotty committed:
add strict typing for iotty

shapournemati-iotty committed:
Merge branch 'iotty' of https://github.com/pburgio/homeassistant.core into iotty

shapournemati-iotty committed:
additional asserts and named snapshots in tests

shapournemati-iotty committed:
fix mypy issues after enabling strict typing

shapournemati-iotty committed:
upgrade iottycloud version to 0.1.3

shapournemati-iotty committed:
move coordinator to runtime_data

shapournemati-iotty committed:
remove entity name

shapournemati-iotty committed:
fix typing issues

emontnemery commented on homeassistant/components/iotty/const.py:
```diff
@@ -0,0 +1,19 @@
+
+IOTTYAPI_BASE = "https://homeassistant.iotty.com/"
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
```
This seems to be unused

shapournemati-iotty replied:
Removed

emontnemery commented on homeassistant/components/iotty/const.py:
```diff
@@ -0,0 +1,19 @@
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+
+LOGGER = logging.getLogger(__package__)
```
It's a bit weird to have both a shared `LOGGER` and individual `_LOGGER` in the modules, pick the pattern you prefer and go for it.

shapournemati-iotty replied:
Done. I have kept the individual `_LOGGER` just fyi

emontnemery commented on homeassistant/components/iotty/const.py:
```diff
@@ -0,0 +1,19 @@
+
+DOMAIN = "iotty"
+
+OAUTH2_AUTHORIZE = "https://auth.iotty.com/.auth/oauth2/login"
+OAUTH2_TOKEN = "https://auth.iotty.com/.auth/oauth2/token"
+OAUTH2_CLIENT_ID = "hass-iotty"
+
+IOTTYAPI_BASE = "https://homeassistant.iotty.com/"
```
These are only used once, move them to `api.py` and `application_credentials.py` respectively please.

shapournemati-iotty replied:
Done

emontnemery commented on homeassistant/components/iotty/__init__.py:
```diff
@@ -0,0 +1,56 @@
+    coordinator: coordinator.IottyDataUpdateCoordinator
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Is it not possible to change like this, here an everywhere else, so mypy knows the type of the config entry runtime data?
```suggestion
async def async_setup_entry(hass: HomeAssistant, entry: IottyConfigEntry) -> bool:
```

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,181 @@
+    """Activate the iotty LightSwitch component."""
+    _LOGGER.debug("Setup SWITCH entry id is %s", config_entry.entry_id)
+
+    coordinator: IottyDataUpdateCoordinator = config_entry.runtime_data.coordinator
```
mypy should be able to infer the type, although it's probably OK to write it out if you think it improves legibility of the code
```suggestion
    coordinator = config_entry.runtime_data.coordinator
```

shapournemati-iotty replied:
Done

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,181 @@
+
+    _attr_has_entity_name = True
+    _attr_name = None
+    _attr_entity_category = EntityCategory.CONFIG
```
I don't think this is correct, it should be `None`?

Documentation here: https://developers.home-assistant.io/docs/core/entity?_highlight=entitycategory#registry-properties

shapournemati-iotty replied:
Yes, having a more thorough look at the documentation and after a couple of local tests I agree with you. It has been changed to `None`

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,181 @@
+        )
+        return self._iotty_device.is_on
+
+    @property
+    def name(self) -> str:
+        """Get the name of this iotty Device."""
+        return self._iotty_device.name
```
This is colliding with the `Entity.name` property, please remove it.

shapournemati-iotty replied:
Done

emontnemery commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,181 @@
+        """Return the device info."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, cast(str, self._attr_unique_id))},
+            name=self.name,
```
```suggestion
            name=self._iotty_device.name,
```

shapournemati-iotty replied:
Done

emontnemery commented on tests/components/iotty/conftest.py:
```diff
@@ -0,0 +1,185 @@
+        unique_id="IOTTY00001",
+    )
+
+    runtime_data = IottyConfigEntryData(set(), MagicMock())
+    mock.runtime_data = runtime_data
```
Why do we mock this? I'd expect the `runtime_data` to be populated when tests configure the config entry?

shapournemati-iotty replied:
I think this is kind of a "leftover" from previous fiddling with testing. It has been removed

emontnemery commented on tests/components/iotty/test_api.py:
```diff
@@ -0,0 +1,79 @@
+) -> None:
+    """Test API creation with no session."""
+
+    with pytest.raises(ValueError) as excinfo:
+        _ = api.IottyProxy(hass, None, None)
+    assert "websession" in str(excinfo.value)
```
Note that you can pass a regex which is matched against the stringified exception: https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.raises
Also, there shouldn't be a need to assign to `_` here?

shapournemati-iotty replied:
I didn't know of that option; implemented like that

emontnemery commented on tests/components/iotty/test_api.py:
```diff
@@ -0,0 +1,79 @@
+
+    iotty = api.IottyProxy(hass, aiohttp_client_session, local_oauth_impl)
+
+    assert iotty is not None
```
This is a meaningless assertion; a newly created Python object can't be `None`. If you just want to assert we can create the object, remove the assignment and assertion, and explain in the test's docstring that we're checking we can create an `IottyProxy` without raising.

shapournemati-iotty replied:
Totally agree! Done

emontnemery commented on tests/components/iotty/test_api.py:
```diff
@@ -0,0 +1,79 @@
+    mock_aioclient: None,
+    aiohttp_client_session: ClientSession,
+) -> None:
+    """Print a message if the token is not valid."""
```
I don't understand the docstring. The test doesn't check if an error message is printed? If you *want* to do that, you can use the caplog fixture.

shapournemati-iotty replied:
No, I think that is an old docstring, I have changed it. Moreover, even though I can see how caplog could be useful, in this scenario I don't find it too valuable to assert on logs. I have added a more specific assertion on the refreshed token instead, which is what we are really testing here

emontnemery commented on tests/components/iotty/test_init.py:
```diff
@@ -0,0 +1,65 @@
+from tests.common import MockConfigEntry
+
+
+async def test_load_unload_coordinator_called(
```
The name of the test, and the use of the `mock_coordinator` fixture suggests we would check the calls to the coordinator, however that doesn't seem to happen?

shapournemati-iotty replied:
Hi, I have added a few assertions on that, let me know if that satisfies you

shapournemati-iotty committed:
coding style fixes

shapournemati-iotty committed:
improve tests coding style and assertion targets

emontnemery commented:
There are a few uncovered lines, mostly related to error handling when we fail to get data from the remote API. Can we add some tests which cover that?

shapournemati-iotty committed:
test edge cases when apis are not working

pburgio committed:
Merge branch 'dev' into iotty

shapournemati-iotty commented on homeassistant/components/iotty/switch.py:
```diff
@@ -0,0 +1,175 @@
+    def device_id(self) -> str:
+        """Get the ID of this iotty Device."""
+        return self._iotty_device.device_id
+
```
Done

shapournemati-iotty commented on tests/components/iotty/test_api.py:
```diff
@@ -0,0 +1,78 @@
+    mock_aioclient: None,
+    aiohttp_client_session: ClientSession,
+) -> None:
+    """A new, valid token is returned if th current token is not valid."""
```
Done

shapournemati-iotty commented on tests/components/iotty/test_api.py:
```diff
@@ -0,0 +1,78 @@
+        hass, DOMAIN, local_oauth_impl
+    )
+
+    _ = api.IottyProxy(hass, aiohttp_client_session, local_oauth_impl)
```
Done

emontnemery commented on tests/components/iotty/test_api.py:
```diff
@@ -0,0 +1,78 @@
+    mock_aioclient: None,
+    aiohttp_client_session: ClientSession,
+) -> None:
+    """If a request with an invalid token is made, a request for a new token is done, and the resulting token is used for future calls."""
```
```suggestion
    """Test getting access token.

    If a request with an invalid token is made, a request for a new token is done,
    and the resulting token is used for future calls.
    """
```

emontnemery commented on tests/components/iotty/test_switch.py:
```diff
@@ -0,0 +1,298 @@
+    async_fire_time_changed(hass)
+    await hass.async_block_till_done()
+
+    # Should have three devices
```
```suggestion
    # Should still have have two devices
```

emontnemery commented on tests/components/iotty/test_switch.py:
```diff
@@ -0,0 +1,298 @@
+
+    # Should have two devices
+    assert hass.states.async_entity_ids_count() == 2
+    assert hass.states.async_entity_ids() == snapshot
```
```suggestion
    entity_ids = assert hass.states.async_entity_ids() == snapshot
    assert entity_ids == snapshot
```

emontnemery commented on tests/components/iotty/test_switch.py:
```diff
@@ -0,0 +1,298 @@
+
+    # Should have three devices
+    assert hass.states.async_entity_ids_count() == 2
+    assert hass.states.async_entity_ids() == snapshot
```
```suggestion
    assert hass.states.async_entity_ids() == entity_ids
```

emontnemery commented on tests/components/iotty/test_switch.py:
```diff
@@ -0,0 +1,298 @@
+    assert hass.states.async_entity_ids() == snapshot
+
+
+async def test_api_throws_response_entities_stay_the_same_as_before(
```
Do the same changes here as in the previous test

shapournemati-iotty committed:
improve tests comments and assertions

emontnemery reviewed (approved):
LGTM, thanks @pburgio , @shapournemati-iotty 👍
